File: examples/langchain-python/background_path.py
import asyncio
import os

from dotenv import load_dotenv
from memexai import MemexAI


async def main() -> None:
    load_dotenv()
    memex = MemexAI(
        url=os.getenv("MEMEX_URL", "http://localhost:8080"),
        api_key=os.getenv("MEMEX_API_KEY", "dev-agent-key"),
        timeout=30,
    )
    try:
        memory = memex.for_user(os.getenv("MEMEX_USER_ID", "langchain_background_user"), actor="langchain-background-path")
        print("background_path: respond first, then a callback/runnable wrapper extracts durable facts")
        insight = "LangChain tool permanent failure: vector store filter syntax does not support nested OR clauses."
        result = await memory.remember({"text": insight, "maxWrites": 2})
        print(result.get("writes", []))
    finally:
        await memex.close()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/openai-sdk-service/src/background-path.ts
import { MemexAI } from "@memexai/sdk"

const memory = new MemexAI({
  url: process.env["MEMEX_URL"] ?? "http://localhost:8080",
  apiKey: process.env["MEMEX_API_KEY"] ?? "dev-agent-key",
}).forUser({
  userId: process.env["MEMEX_USER_ID"] ?? "openai_sdk_background_user",
  actor: "openai-sdk-background-path",
})

console.log("background_path: complete the user response, then learn from app tool results")
console.log("assistant: I found the SQL incompatibility and adjusted the query.")

const toolResult = {
  toolName: "run_sql",
  ok: false,
  permanent: true,
  error: "ILIKE is not supported in this SQL dialect; use LOWER(column) LIKE LOWER(pattern).",
}

if (!toolResult.ok && toolResult.permanent) {
  const insight = `${toolResult.toolName} permanent failure: ${toolResult.error}`
  const result = await memory.remember({ text: insight, toolCallId: "openai_background_run_sql" })
  console.log(`learned_after_response: ${insight}`)
  console.log(`writes: ${result.writes.map((write) => write.path).join(", ") || "none"}`)
}
